modules/local/emptydrops.nf

```diff
@@ -0,0 +1,82 @@
+process EMPTYDROPS_CELL_CALLING {
+    tag "$meta.id"
+    label 'process_medium'
+
+    conda "bioconda::bioconductor-dropletutils"
+    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
+        'https://depot.galaxyproject.org/singularity/bioconductor-dropletutils:1.18.0--r42hf17093f_1' :
+        'quay.io/biocontainers/bioconductor-dropletutils:1.18.0--r42hf17093f_1' }"
+
+    input:
+    // inputs from cellranger nf-core module does not come in a single sample dir
+    // for each sample, the sub-folders and files come directly in array.
+    tuple val(meta), path(inputs)
+
+    output:
+    tuple val(meta), path("emptydrops_filtered"), emit: filtered_matrices
+
+    when:
+    task.ext.when == null || task.ext.when
+
+    script:
+    if (params.aligner == "cellranger") {
+
+        matrix   = "matrix.mtx.gz"
+        barcodes = "barcodes.tsv.gz"
+        features = "features.tsv.gz"
+
+    } else if (params.aligner == "kallisto") {
+
+        matrix     = "counts_unfiltered/*.mtx"
+        barcodes   = "counts_unfiltered/*.barcodes.txt"
+        features   = "counts_unfiltered/*.genes.txt"
+
+    } else if (params.aligner == "alevin") {
+
+        matrix   = "*_alevin_results/af_quant/alevin/quants_mat.mtx"
+        barcodes = "*_alevin_results/af_quant/alevin/quants_mat_rows.txt"
+        features = "*_alevin_results/af_quant/alevin/quants_mat_cols.txt"
+
+    } else if (params.aligner == 'star') {
+
+        matrix   = "raw/matrix.mtx.gz"
+        barcodes = "raw/barcodes.tsv.gz"
+        features = "raw/features.tsv.gz"
+
+    }
+
+    //
+    // run script
+    //
+    if (params.aligner == 'kallisto' && params.kb_workflow != 'standard')
+    """
+    mkdir emptydrops_filtered/
+    # convert file types
+    for splice_type in spliced unspliced ; do
+        emptydrops_cell_calling.R \\
+            counts_unfiltered/\${splice_type}.mtx \\
+            counts_unfiltered/\${splice_type}.barcodes.txt \\
+            counts_unfiltered/\${splice_type}.genes.txt \\
+            emptydrops_filtered \\
+            ${params.aligner} \\
+            0
+    done
+    """
```

Comment:
@gennadyFauna will this still work with the latest kallisto update?

Reply:
For more context, this PR adds emptyDrops as an external tool for detecting empty droplets that can be applied to all aligners. I think you reworked the `kb_workflows` in the kallisto PR, so I was wondering if anything needs to be changed here in preparation of getting #294 merged.

Reply:
@grst It will not. `spliced` and `unspliced` are only present in the `lamanno` workflow (which is obsolete since 0.28.0). Also, in `nac`, the convention for mtx files is different: nascent, mature, and ambiguous all share the same `genes` and `barcodes` files.